usersession/userd/privileged_desktop_launcher_internal_test.go

```diff
@@ -0,0 +1,486 @@
+		"mir-kiosk-scummvm-mir-kiosk-scummvm.desktop",
+		"bar-foo-baz.desktop",
+		"bar-baz-foo.desktop",
+		"foo-bar_foo-bar.desktop",
+	}
+
+	for _, id := range desktopIdTests {
+		_, err := userd.DesktopFileIDToFilename(id)
+		c.Assert(err, NotNil)
+	}
+}
+
+func (s *privilegedDesktopLauncherInternalSuite) TestParseExecCommandSucceedsWithValidEntry(c *C) {
+	var exec_command = []struct {
+		exec_command string
+		expect       []string
+	}{
+		// valid with no exec variables
+		{"env BAMF_DESKTOP_FILE_HINT=/var/lib/snapd/desktop/applications/mir-kiosk-scummvm_mir-kiosk-scummvm.desktop /snap/bin/mir-kiosk-scummvm %U",
+			[]string{"env", "BAMF_DESKTOP_FILE_HINT=/var/lib/snapd/desktop/applications/mir-kiosk-scummvm_mir-kiosk-scummvm.desktop", "/snap/bin/mir-kiosk-scummvm"}},
+		// valid with literal '%' and no exec variables
+		{"/snap/bin/foo -f %%bar", []string{"/snap/bin/foo", "-f", "%bar"}},
+		{"/snap/bin/foo -f %%bar %%baz", []string{"/snap/bin/foo", "-f", "%bar", "%baz"}},
+		// valid where quoted strings are passed through
+		{"/snap/bin/foo '-f %U'", []string{"/snap/bin/foo", "-f %U"}},
+		{"/snap/bin/foo '-f %%bar'", []string{"/snap/bin/foo", "-f %%bar"}},
+		{"/snap/bin/foo '-f %U %%bar'", []string{"/snap/bin/foo", "-f %U %%bar"}},
+		{"/snap/bin/foo \"'-f bar'\"", []string{"/snap/bin/foo", "'-f bar'"}},
+		{"/snap/bin/foo '\"-f bar\"'", []string{"/snap/bin/foo", "\"-f bar\""}},
+		// valid with exec variables stripped out
+		{"/snap/bin/foo -f %U", []string{"/snap/bin/foo", "-f"}},
+		{"/snap/bin/foo -f %U %i", []string{"/snap/bin/foo", "-f", "--icon", "/snap/chromium/1193/chromium.png"}},
+		{"/snap/bin/foo -f %U bar", []string{"/snap/bin/foo", "-f", "bar"}},
+		{"/snap/bin/foo -f %U bar %i", []string{"/snap/bin/foo", "-f", "bar", "--icon", "/snap/chromium/1193/chromium.png"}},
+		// valid with mixture of literal '%' and exec variables
+		{"/snap/bin/foo -f %U %%bar", []string{"/snap/bin/foo", "-f", "%bar"}},
+		{"/snap/bin/foo -f %U %i %%bar", []string{"/snap/bin/foo", "-f", "--icon", "/snap/chromium/1193/chromium.png", "%bar"}},
+		{"/snap/bin/foo -f %U %%bar %i", []string{"/snap/bin/foo", "-f", "%bar", "--icon", "/snap/chromium/1193/chromium.png"}},
+		{"/snap/bin/foo -f %%bar %U %i", []string{"/snap/bin/foo", "-f", "%bar", "--icon", "/snap/chromium/1193/chromium.png"}},
+	}
+
+	for _, test := range exec_command {
+		actual, err := userd.ParseExecCommand(test.exec_command, "/snap/chromium/1193/chromium.png")
+		c.Assert(err, IsNil)
+		c.Assert(actual, DeepEquals, test.expect)
+	}
+}
+
+func (s *privilegedDesktopLauncherInternalSuite) TestParseExecCommandFailsWithInvalidEntry(c *C) {
+	// the only invalid entries are those that error from shlex.Split()
+	var exec_command = []string{
+		"/snap/bin/foo \"unclosed double quote",
+		"/snap/bin/foo 'unclosed single quote",
+	}
+
+	for _, test := range exec_command {
+		_, err := userd.ParseExecCommand(test, "/snap/chromium/1193/chromium.png")
+		c.Assert(err, NotNil)
```

Comment:
as a pattern, I think we prefer to use `ErrorMatches` rather than NotNil so that the reader of the test can see what the expected returned error is, for many validate functions, we will sometimes see different errors for different validation errors, and sometimes when we refactor validation functions, we would expect tests to fail based on the changes we made, but they don't because the test isn't checking the error messages, it's just checking that there was an error and this leads to subtle assumptions by the developer about the returned error messages. An example would be my recent misadventures in the gadget package, where I refactored something and the error message being returned suddenly changed which made me realize that the order I was checking things with was wrong, and that I needed to refactor my refactoring. If that test was only checking that the error was non-nil I would not have realized the issue with my implementation and would have introduced a subtle bug.